### PR TITLE
Use a more general approach

### PR DESCRIPTION
**Pull request by bemasc**
This design can handle ECH configurations in a wider variety of architectures, and is extensible to future origin configuration needs as well.

### CODE REVIEW

**Comment by bemasc** [comment]
@sftcd @richsalz IIRC we agreed to try out a design that could represent aliases, so here you go.  I know this is a big change with two weekdays until the draft deadline, so it's OK if we can't get it out in time, but I figured I would try.

File: draft-ietf-tls-wkech.txt [hunk 1]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
-   o  A cronjob creates creates a JSON file for each backend site at
-      https://$FRONT/.well-known/ech/$BACKEND.json
-   o  Each JSON file contains an array with the ECHConfigList values
-      values for that particular $BACKEND as shown in Figure 1 - the
-      values in Figure 1 with ellipses are the values we want to
-      eventually see in the DNS
+   o  A cronjob creates creates a JSON file containing the ECHConfigList
+      site at https://$ORIGIN/.well-known/dns-zone-info
+   o  The JSON file contains an array with the ECHConfigList values
+      values as shown in Figure 1 - the values in Figure 1 with ellipses
```
Comment (richsalz) [inline draft-ietf-tls-wkech.txt:145]: nit: IIRC, the values have ".." which is one dot short of an ellipsis.

File: draft-ietf-tls-wkech.txt [hunk 2]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
-   The JSON file contains an array for a couple of reasons:
+   o  priority: The value is a positive integer corresponding to the
+      SvcPriority.  If omitted, the zone factory SHOULD infer increasing
+      SvcPriority from the order of the endpoints array.
+   o  target: The value is a fully qualified domain name, corresponding
+      to the HTTPS record's TargetName.  If omitted, the zone factory
+      SHOULD infer a value of ".".
+   o  port: The value is a non-negative integer, corresponding to the
+      value of the "port" SvcParamKey.
+   o  ech: The value is an ECHConfigList as a base64 encoded string,
```
Comment (richsalz) [inline draft-ietf-tls-wkech.txt:212]: s/as a/encoded as a/

File: draft-ietf-tls-wkech.txt [hunk 3]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
-   the latest version of the JSON file.  This leaves the control of
-   "expiry" with the web server, so long as the ECHConfigList values
-   presented actually work.  [[An alternative could be to have the new
-   values just be appended to the zone, but that'd require some form of
-   "notAfter" value in the JSON file which seems unnecessary and more
-   complex.]]
+   The zone factory MUST set a DNS TTL short enough that any generated
+   records expire from DNS caches before the JSON object's HTTP cache
+   lifetime expires.  The zone factory MUST refresh the JSON object and
+   regenerate the zone before it expires each time.  This ensures that
```
Comment (richsalz) [inline draft-ietf-tls-wkech.txt:272]: nice!

File: draft-ietf-tls-wkech.txt [hunk 4]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
 
 
-Farrell, et al.          Expires April 17, 2023                 [Page 5]
+Farrell, et al.          Expires April 23, 2023                 [Page 5]
 

 Internet-Draft           Well-Known URI for ECH             October 2022
 
 
+   A possibly interesting (unintended) consequence of this design is
```
Comment (richsalz) [inline draft-ietf-tls-wkech.txt:285]: Well, this whole paragraph is kinda casual and cutesy. Hard to see justification for unintended tho.  And not sure "fun" is appropriate.

**Review by richsalz — CHANGES_REQUESTED** [review]
I'm not sure that expanding this to full ZONE information is appropriate, as opposed to limiting to SVCB things, but I am fine with the WG and DNS folks making that decision.

File: draft-ietf-tls-wkech.txt [hunk 5]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
-   defined in Section 5 at: https://$FRONT/.well-known/ech/$BACKEND.json
+   If a web server ($ORIGIN) wants to convey information to the Zone
+   Factory, it publishes the JSON content defined in Section 5 at:
+   https://$ORIGIN/.well-known/dns-zone-info
 
    The well-known URI defined here MUST be an https URL and therefore
-   the zone factory verifies the correct $FRONT is being accessed.  If
+   the zone factory verifies the correct $ORIGIN is being accessed.  If
    there is any failure in accessing the well-known URI, then the zone
    factory MUST NOT modify the zone.
```
[the remark this answers is not in the document]
Comment (bemasc) [reply draft-ietf-tls-wkech.txt:165]: I'm really not sure what to do in that case.  I added a TODO.

File: draft-ietf-tls-wkech.txt (hunk 4 above)
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: OK, I've rewritten this in formal terms and moved it to the Security Considerations.

File: draft-ietf-tls-wkech.txt [hunk 6]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
-   work with the $BACKEND server before publication.  A "special" TLS
-   client may be needed for this check, that does not require the
-   ECHConfigList value to have already been published in the DNS.  [[I
-   guess that calls for the zone factory to know of a "safe" URL on
-   $BACKEND to try, or maybe it could use HTTP HEAD?  Figuring that out
-   is TBD.  The ZF could also try a GREASEd ECH and see if the retry-
-   configs it gets back is one of the ECHConfig values in the
-   ECHConfigList.]]
+   The zone factory SHOULD check that the presented endpoints work and
+   provide access to $ORIGIN before publication.  A "special" TLS client
```
[the remark this answers is not in the document]
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: Done

File: draft-ietf-tls-wkech.txt (hunk 2 above)
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: Rephrased

File: draft-ietf-tls-wkech.txt [hunk 7]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
+   The JSON file at the well-known URI MUST contain an object with
+   either an "endpoints" key or an "alias" key.  If the "endpoints" key
+   is present, its value is an array whose elements represent HTTPS
+   records in ServiceMode.  Each element MAY contain one or more keys
+   from the JSON HTTP Origin Info registry (see IANA Considerations).
+   The initial registry entries are:
 
-   The JSON file contains an array for a couple of reasons:
+   o  priority: The value is a positive integer corresponding to the
+      SvcPriority.  If omitted, the zone factory SHOULD infer increasing
```
[the remark this answers is not in the document]
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: With SvcPriority, lower is better.  This has created countless wording problems.  Changed from "increasing" to "numerically increasing".

File: draft-ietf-tls-wkech.txt [hunk 8]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
-   o  Web server has a "front-end" site ($FRONT), where $FRONT will
-      typically be the DNS name used in the ECHConfigList public_name
-      field for ECHConfig version 0xff0d
-   o  A cronjob creates creates a JSON file for each backend site at
-      https://$FRONT/.well-known/ech/$BACKEND.json
-   o  Each JSON file contains an array with the ECHConfigList values
-      values for that particular $BACKEND as shown in Figure 1 - the
-      values in Figure 1 with ellipses are the values we want to
-      eventually see in the DNS
+   o  A cronjob creates creates a JSON file containing the ECHConfigList
```
[the remark this answers is not in the document]
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: Done

File: draft-ietf-tls-wkech.txt (hunk 1 above)
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: I've rewritten this list to reflect the cache-lifetime-driven behavior in this PR.

File: draft-ietf-tls-wkech.txt [hunk 9]
```diff
@@ -131,157 +139,155 @@ Internet-Draft           Well-Known URI for ECH             October 2022
...
+      values as shown in Figure 1 - the values in Figure 1 with ellipses
+      are the values we want to eventually see in the DNS
    o  On the zone factory, a cronjob runs at N+3 past the hour, it knows
       all the names involved and checks to see if the content at those
       well-known URIs has changed or not
    o  If the content has changed the cronjob attempts to use the
-      ECHConfigList values, and for each $BACKEND where that works, it
-      updates the zone file and re-publishes the zone containing only
-      the new ECHConfigList values
+      ECHConfigList values, and if that works, it updates the zone file
```
[the remark this answers is not in the document]
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: OK, rephrased.

**Review by bemasc — COMMENTED** [review]
> I'm not sure that expanding this to full ZONE information is appropriate, as opposed to limiting to SVCB things,

OK, I've changed the focus to be more specifically SVCB, especially the abstract and the name of the .well-known resource.

File: draft-ietf-tls-wkech.txt [hunk 10]
```diff
@@ -5,19 +5,22 @@
...
 
          A well-known URI for publishing ECHConfigList values.
                         draft-ietf-tls-wkech-00
 
 Abstract
 
-   We propose use of a well-known URI at which web servers can publish
-   ECHConfigList values as a way to help get those published in the DNS.
+   We propose use of a well-known URI at which an HTTP origin can inform
+   its authoritative DNS server about information that should appear in
```
[the remark this answers is not in the document]
Comment (bemasc) [reply draft-ietf-tls-wkech.txt]: Done

**Comment by sftcd** [comment]
I'm not sure whether or not I like that approach, however it
is entirely fair to ask the WG if they prefer the highly
specific one from -00 or this more generic one in -01.

I'd say just add in some text like that, e.g. maybe "This
version proposes changing from the highly ECHConfig specific
approach of -00 to a much more generic approach - the authors
are including this as a way to get feedback from the WG as
to which of these approaches may be more likely to garner
rough consensus. (IOW) If the WG feel this is worse than
-00 we're fine with reverting."

Maybe whack something like that into the abstract or the
early intro so's it's noticeable?

Cheers,
S.

**Comment by sftcd** [comment]
On 20/10/2022 22:12, Benjamin M. Schwartz wrote:
> can inform
> +   its authoritative DNS server

Is "it's" really correct? Not in the original use case where
the web origin is the public name talking to the ZF of the
backend. I see that's better nuanced in the text but maybe
s/it's/an/

S.

**Comment by bemasc** [comment]
> I'd say just add in some text like that, e.g. maybe "This version proposes changing from the highly ECHConfig specific approach of -00 to a much more generic approach - the authors are including this as a way to get feedback from the WG as to which of these approaches may be more likely to garner rough consensus. (IOW) If the WG feel this is worse than -00 we're fine with reverting."

Done

> maybe s/it's/an/

Done

**Review by richsalz — APPROVED** [review]
lgtm.

**Comment by sftcd** [comment]
On 20/10/2022 22:23, Benjamin M. Schwartz wrote:
>> I'd say just add in some text like that, e.g. maybe "This version
>> proposes changing from the highly ECHConfig specific approach of
>> -00 to a much more generic approach - the authors are including
>> this as a way to get feedback from the WG as to which of these
>> approaches may be more likely to garner rough consensus. (IOW) If
>> the WG feel this is worse than -00 we're fine with reverting."
> 
> Done

Cool. With that included, I'm fine with hitting the publish
button if you want. If I need to do it 'cause I'm adding you
folks I can do it tomorrow or over the w/e.

S

> 
>> maybe s/it's/an/
> 
> Done
>

**Pull request merged**